File: android/src/main/kotlin/com/ovaom/flutter_libpd/pd_controller/pd_messenger/PdMessenger.kt
package com.ovaom.flutter_libpd.pd_controller.pd_messenger

interface PdMessenger {
    fun sendFloat(message: FloatMessage)
    fun sendBang(message: BangMessage)
    fun sendSymbol(message: SymbolMessage)
    fun sendList(message: ListMessage)
    fun sendMessage(message: ArgumentMessage)
}

sealed class MessageError : Exception() {
    object MessageFailed: MessageError()
}

data class FloatMessage(val value: Float, val receiver: String)
data class BangMessage(val receiver: String)
data class SymbolMessage(val value: String, val receiver: String)
data class ListMessage(val value: List<Any>, val receiver: String)
data class ArgumentMessage(val value: String, val arguments: List<Any>, val receiver: String)

File: ios/Classes/PdController/PdMessenger/PdMessenger.swift
//
//  PdMessenger.swift
//  Runner
//
//  Created by flav on 23/05/2022.
//

import Foundation

public protocol PdMessenger {
    func sendFloat(_ message: FloatMessage) throws
    func sendBang(_ message: BangMessage) throws
    func sendSymbol(_ message: SymbolMessage) throws
    func sendList(_ message: ListMessage) throws
    func sendMessage(_ message: ArgumentMessage) throws
}

public enum MessageError: Error {
    case messageFailed
}

public struct FloatMessage {
    public let value: Float
    public let receiver: String
    
    public init(value: Float, receiver: String) {
        self.value = value
        self.receiver = receiver
    }
}

public struct BangMessage {
    public let receiver: String

    public init(receiver: String) {
        self.receiver = receiver
    }
}

public struct SymbolMessage {
    public let value: String
    public let receiver: String
    
    public init(value: String, receiver: String) {
        self.value = value
        self.receiver = receiver
    }
}

public struct ListMessage {
    public let value: [Any]
    public let receiver: String
    
    public init(value: [Any], receiver: String) {
        self.value = value
        self.receiver = receiver
    }
}

public struct ArgumentMessage {
    public let value: String
    public let arguments: [Any]
    public let receiver: String
    
    public init(value: String, arguments: [Any], receiver: String) {
        self.value = value
        self.arguments = arguments
        self.receiver = receiver
    }
}
